cryptography/src/bls12381/primitives/ops/aggregate.rs

```diff
@@ -0,0 +1,490 @@
+/// This type is returned by [`combine_signatures`] and ensures that
+/// aggregated signatures are not confused with individual signatures.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Signature<V: Variant>(V::Signature);
+
+impl<V: Variant> Signature<V> {
+    /// Returns the inner signature value.
+    pub const fn inner(&self) -> &V::Signature {
+        &self.0
+    }
+
+    /// Creates a zero aggregate signature.
+    pub fn zero() -> Self {
+        Self(V::Signature::zero())
+    }
+}
+
+impl<V: Variant> Write for Signature<V> {
+    fn write(&self, writer: &mut impl BufMut) {
+        self.0.write(writer);
+    }
+}
+
+impl<V: Variant> Read for Signature<V> {
+    type Cfg = ();
+
+    fn read_cfg(reader: &mut impl Buf, _cfg: &Self::Cfg) -> Result<Self, CodecError> {
+        Ok(Self(V::Signature::read(reader)?))
+    }
+}
+
+impl<V: Variant> FixedSize for Signature<V> {
+    const SIZE: usize = V::Signature::SIZE;
+}
+
+#[cfg(feature = "arbitrary")]
+impl<V: Variant> arbitrary::Arbitrary<'_> for Signature<V>
+where
+    V::Signature: for<'a> arbitrary::Arbitrary<'a>,
+{
+    fn arbitrary(u: &mut arbitrary::Unstructured<'_>) -> arbitrary::Result<Self> {
+        Ok(Self(V::Signature::arbitrary(u)?))
+    }
+}
+
+/// Combines multiple public keys into an aggregate public key.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on all `public_keys`,
+/// that each `public_key` is unique, and that the caller has a Proof-of-Possession (PoP)
+/// for each `public_key`. If any of these assumptions are violated, an attacker can
+/// exploit this function to verify an incorrect aggregate signature.
+pub fn combine_public_keys<'a, V, I>(public_keys: I) -> PublicKey<V>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Public>,
+    V::Public: 'a,
+{
+    let mut p = V::Public::zero();
+    for pk in public_keys {
+        p += pk;
+    }
+    PublicKey(p)
+}
+
+/// Combines multiple signatures into an aggregate signature.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on each `signature` and
+/// that each `signature` is unique. If any of these assumptions are violated, an attacker can
+/// exploit this function to verify an incorrect aggregate signature.
+pub fn combine_signatures<'a, V, I>(signatures: I) -> Signature<V>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Signature>,
+    V::Signature: 'a,
+{
+    let mut s = V::Signature::zero();
+    for sig in signatures {
+        s += sig;
+    }
+    Signature(s)
+}
+
+/// Verifies the aggregate signature over a single message from multiple public keys.
+///
+/// # Warning
+///
+/// This function assumes the caller has performed a group check and collected a proof-of-possession
+/// for all provided `public`. This function assumes a group check was already performed on the
+/// `signature`. It is not safe to provide duplicate public keys.
+pub fn verify_multiple_public_keys<'a, V, I>(
```

Comment:
```suggestion
pub fn verify<'a, V, I>(
```

Given that this is namespaced, we can use a much simpler name.

Reply:
A more advanced idea:

Have the `PublicKey` type in this module (an aggregate public key) implement the `PublicKey` trait, with an associated type of `Signature` (an aggregate signature)